osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreProcessor.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Rulesets.Osu.Difficulty
+{
+    internal class OsuLegacyScoreProcessor : ILegacyScoreProcessor
+    {
+        public int AccuracyScore { get; private set; }
+
+        public int ComboScore { get; private set; }
+
+        public double BonusScoreRatio => legacyBonusScore == 0 ? 0 : (double)modernBonusScore / legacyBonusScore;
+
+        private int legacyBonusScore;
+        private int modernBonusScore;
+        private int combo;
+
+        private double scoreMultiplier;
+        private IBeatmap playableBeatmap = null!;
+
+        public void Simulate(IWorkingBeatmap workingBeatmap, IBeatmap playableBeatmap, IReadOnlyList<Mod> mods)
+        {
+            this.playableBeatmap = playableBeatmap;
+
+            IBeatmap baseBeatmap = workingBeatmap.Beatmap;
+
+            int countNormal = 0;
+            int countSlider = 0;
+            int countSpinner = 0;
+
+            foreach (HitObject obj in workingBeatmap.Beatmap.HitObjects)
+            {
+                switch (obj)
+                {
+                    case IHasPath:
+                        countSlider++;
+                        break;
+
+                    case IHasDuration:
+                        countSpinner++;
+                        break;
+
+                    default:
+                        countNormal++;
+                        break;
+                }
+            }
+
+            int objectCount = countNormal + countSlider + countSpinner;
+
+            int drainLength = 0;
+
+            if (baseBeatmap.HitObjects.Count > 0)
+            {
+                int breakLength = baseBeatmap.Breaks.Select(b => (int)Math.Round(b.EndTime) - (int)Math.Round(b.StartTime)).Sum();
+                drainLength = ((int)Math.Round(baseBeatmap.HitObjects[^1].StartTime) - (int)Math.Round(baseBeatmap.HitObjects[0].StartTime) - breakLength) / 1000;
```

Comment:
The last object in the beatmap may not be necessarily the one with the greatest end time. I'm not sure this is necessarily going to matter because the only ruleset where it definitely would (mania) doesn't compute this. But it might, on some freak beatmaps.

Reply:
In fact we have a helper for this:

https://github.com/ppy/osu/blob/81e759cb6b18b82b9cd4716a35ad8ce5b2a49b11/osu.Game/Beatmaps/IBeatmap.cs#L144-L156

Reply:
This is intentional, actually.

I'm not sure if you have access to the stable reference source, but here's the corresponding lines: https://github.com/peppy/osu-stable-reference/blob/1531237b63392e82c003c712faa028406073aa8f/osu!/GameplayElements/Beatmaps/Beatmap.cs#L627-L650

Notice how for `countSpinner`, it sets `lastTimeStr = split[2]` which is corresponds to the `time` (start time) parameter of the slider https://osu.ppy.sh/wiki/en/Client/File_formats/osu_(file_format)#sliders

Reply:
yikes, okay :skull: